pyiron_base/storage/helper_functions.py

```diff
@@ -40,3 +104,218 @@ def write_hdf5(
+                              {
+                                  'hdf5root/group/node_name': {},
+                                  'hdf5root/group/subgroup/node_name': [...],
+                              }
+        compression (int): Compression level to use (0-9) to compress data using gzip.
+        slash (str): 'error' | 'replace' Whether to replace forward-slashes ('/') in any key found nested within
+                      keys in data. This does not apply to the top level name (title). If 'error', '/' is not allowed
+                      in any lower-level keys.
+    """
+    with open_hdf5(file_name, mode="a") as store:
+        for k, v in data_dict.items():
+            write_hdf5_with_json_support(
+                file_handle=store,
+                value=v,
+                path=get_h5_path(h5_path=h5_path, name=k),
+                compression=compression,
+                slash=slash,
+            )
+
+
+def list_groups_and_nodes(hdf, h5_path):
+    """
+    Get the list of groups and list of nodes from an open HDF5 file
+
+    Args:
+        hdf (h5py.File): file handle of an open HDF5 file
+        h5_path (str): path inside the HDF5 file
+
+    Returns:
+        list, list: list of groups and list of nodes
+    """
+    groups = set()
+    nodes = set()
+    try:
+        h = hdf[h5_path]
+        for k in h.keys():
+            if isinstance(h[k], h5py.Group):
+                groups.add(k)
+            else:
+                nodes.add(k)
+    except KeyError:
+        pass
+    return list(groups), list(nodes)
+
+
+def read_dict_from_hdf(file_name, h5_path, group_paths=[], slash="ignore"):
+    """
+    Read data from HDF5 file into a dictionary - by default only the nodes are converted to dictionaries, additional
+    groups can be specified using the group_paths parameter.
+
+    Args:
+       hdf (pyiron_base.storage.hdfio.FileHDFio): HDF5 file object
+       file_name (str): Name of the file on disk
+       h5_path (str): Path to a group in the HDF5 file where the data is stored
+       group_paths (list): list of additional groups to be included in the dictionary, for example:
+                           ["input", "output", "output/generic"]
+                           These groups are defined relative to the h5_path.
+       slash (str): 'ignore' | 'replace' Whether to replace the string {FWDSLASH} with the value /. This does
+                    not apply to the top level name (title). If 'ignore', nothing will be replaced.
+    Returns:
+       dict:     The loaded data. Can be of any type supported by ``write_hdf5``.
+    """
+
+    def get_dict_from_nodes(store, h5_path, slash="ignore"):
+        return {
+            n: read_hdf5(
+                fname=store, title=get_h5_path(h5_path=h5_path, name=n), slash=slash
+            )
+            for n in list_groups_and_nodes(hdf=store, h5_path=h5_path)[1]
+        }
+
+    def resolve_nested_dict(group_path, data_dict):
```

Comment:
```suggestion
    def resolve_nested_dict(group_path, data_dict):
        """Turns a dict with a key containing slashes into a nested dict.  {'/a/b/c': 1} -> {'a': {'b': {'c': 1}"""
```

Reply:
Actually maybe the whole body can be just
```python
groups = group_path.split("/")
nested_dict = data_dict
for g in groups[::-1]:
  nested_dict = {g: nested_dict}
```
avoid the recursion which I found a bit confusing.